pkg/executor/sortexec/topn.go

```diff
@@ -166,76 +251,317 @@ func (e *TopNExec) loadChunksUntilTotalLimit(ctx context.Context) error {
 			break
 		}
 		e.chkHeap.rowChunks.Add(srcChk)
+		if e.spillHelper.isSpillNeeded() {
+			e.isSpillTriggeredInStage1ForTest = true
+			break
+		}
+
+		injectTopNRandomFail(1)
 	}
-	e.initPointers()
-	e.initCompareFuncs()
-	e.buildKeyColumns()
+
+	e.chkHeap.initPtrs()
 	return nil
 }
 
 const topNCompactionFactor = 4
 
-func (e *TopNExec) executeTopN(ctx context.Context) error {
-	heap.Init(e.chkHeap)
-	for uint64(len(e.chkHeap.rowPtrs)) > e.totalLimit {
-		// The number of rows we loaded may exceeds total limit, remove greatest rows by Pop.
-		heap.Pop(e.chkHeap)
+func (e *TopNExec) executeTopNNoSpill(ctx context.Context) error {
+	if e.spillHelper.isSpillNeeded() {
+		e.isSpillTriggeredInStage2ForTest = true
+		return nil
 	}
+
 	childRowChk := exec.TryNewCacheChunk(e.Children(0))
 	for {
+		if e.spillHelper.isSpillNeeded() {
+			e.isSpillTriggeredInStage2ForTest = true
+			return nil
+		}
+
 		err := exec.Next(ctx, e.Children(0), childRowChk)
 		if err != nil {
 			return err
 		}
+
 		if childRowChk.NumRows() == 0 {
 			break
 		}
-		err = e.processChildChk(childRowChk)
-		if err != nil {
-			return err
-		}
+
+		e.chkHeap.processChkNoSpill(childRowChk)
+
 		if e.chkHeap.rowChunks.Len() > len(e.chkHeap.rowPtrs)*topNCompactionFactor {
-			err = e.doCompaction(e.chkHeap)
+			err = e.chkHeap.doCompaction(e)
 			if err != nil {
 				return err
 			}
 		}
+		injectTopNRandomFail(10)
+	}
+
+	slices.SortFunc(e.chkHeap.rowPtrs, e.chkHeap.keyColumnsCompare)
+	return nil
+}
+
+func (e *TopNExec) spillRemainingRowsWhenNeeded() error {
+	if e.spillHelper.isSpillTriggered() {
+		return e.spillHelper.spill()
+	}
+	return nil
+}
+
+func (e *TopNExec) checkSpillAndExecute() error {
+	if e.spillHelper.isSpillNeeded() {
+		// Wait for the stop of all workers
+		e.fetcherAndWorkerSyncer.Wait()
+		return e.spillHelper.spill()
 	}
-	slices.SortFunc(e.chkHeap.rowPtrs, e.keyColumnsCompare)
 	return nil
 }
 
-func (e *TopNExec) processChildChk(childRowChk *chunk.Chunk) error {
-	for i := 0; i < childRowChk.NumRows(); i++ {
-		heapMaxPtr := e.chkHeap.rowPtrs[0]
-		var heapMax, next chunk.Row
-		heapMax = e.chkHeap.rowChunks.GetRow(heapMaxPtr)
-		next = childRowChk.GetRow(i)
-		if e.chkHeap.greaterRow(heapMax, next) {
-			// Evict heap max, keep the next row.
-			e.chkHeap.rowPtrs[0] = e.chkHeap.rowChunks.AppendRow(childRowChk.GetRow(i))
-			heap.Fix(e.chkHeap, 0)
+func (e *TopNExec) fetchChunksFromChild(ctx context.Context) {
+	defer func() {
+		if r := recover(); r != nil {
+			processPanicAndLog(e.resultChannel, r)
 		}
+
+		e.fetcherAndWorkerSyncer.Wait()
+		err := e.spillRemainingRowsWhenNeeded()
+		if err != nil {
+			e.resultChannel <- rowWithError{err: err}
+		}
+
+		close(e.chunkChannel)
+	}()
+
+	for {
+		chk := exec.TryNewCacheChunk(e.Children(0))
+		err := exec.Next(ctx, e.Children(0), chk)
+		if err != nil {
+			e.resultChannel <- rowWithError{err: err}
+			return
+		}
+
+		rowCount := chk.NumRows()
+		if rowCount == 0 {
+			break
+		}
```

Comment:
If all chunks are read and `fetchChunksFromChild` exits, who triggers spill?

Reply:
> If all chunks are read and `fetchChunksFromChild` exits, who triggers spill?

Function `generateTopNResultsWithNoSpill` will check if spill flag is set and spill will be trigger the spill in function `generateTopNResultsWithSpill`.

Reply:
ok